### PR TITLE
Drop `replacebranch` in favor of `parseRoute`

### PR DESCRIPTION
**Pull request by yakov116**
1. LINKED ISSUES:
   Follow up to #3083

2. TEST URLS:
	### Default branch button
	https://github.com/sindresorhus/refined-github/commits/incremental-tag-changelog-link
	With pagnation
   https://github.com/sindresorhus/refined-github/commits/incremental-tag-changelog-link?after=51975964dd72a947a51b9a7be118d75e971549d1+34
	#### On a file
	Default branch: https://github.com/sindresorhus/refined-github/blob/master/.editorconfig
	https://github.com/sindresorhus/refined-github/blob/incremental-tag-changelog-link/.editorconfig

	### Latest tag button
	https://github.com/sindresorhus/refined-github/tree/incremental-tag-changelog-link
	https://github.com/sindresorhus/refined-github
	#### On a file
	Default branch: https://github.com/sindresorhus/refined-github/blob/master/.editorconfig
	https://github.com/sindresorhus/refined-github/blob/incremental-tag-changelog-link/.editorconfig

3. SCREENSHOT:
   None

### CODE REVIEW

**Comment by fregante** [comment]
parseRoute should not change. The operation is identical to what other features do.

**Comment by yakov116** [comment]
@fregante OK
All the other features needed to change the route or something more. Here its the same function just replacing the current branch with something else.

I will change it.

File: source/libs/utils.ts [hunk 1]
```diff
@@ -173,7 +164,7 @@ export function parseRoute(pathname: string): string[] {
 		throw new Error('The branch of the current page must match the branch in the `pathname` parameter');
 	}
 
-	const filePath = parts.replace(currentBranch + '/', '');
+	const filePath = parts.replace(currentBranch, '').replace(/^\//, '');
```
Comment (yakov116) [inline source/libs/utils.ts:167]: @fregante this change is still needed, there may not be a leading slash.

File: source/features/default-branch-button.tsx [hunk 2]
```diff
@@ -17,11 +17,18 @@ async function init(): Promise<false | void> {
...
+		return false;
+	}
+
 	let url;
 	if (pageDetect.isRepoRoot()) {
 		url = `/${getRepoURL()}`;
 	} else {
-		url = replaceBranch(currentBranch, defaultBranch);
+		pathParts[5] = defaultBranch;
+		url = pathParts.join('/') + location.search;
```
Comment (fregante) [inline source/features/default-branch-button.tsx:31]: No need to preserve the search

File: source/features/latest-tag-button.tsx [hunk 3]
```diff
@@ -95,7 +95,9 @@ async function init(): Promise<false | void> {
 	if (pageDetect.isRepoRoot()) {
 		href = `/${getRepoURL()}/tree/${latestTag}`;
 	} else {
-		href = replaceBranch(currentBranch, latestTag);
+		const pathParts = parseRoute(location.pathname);
+		pathParts[5] = latestTag;
```
Comment (fregante) [inline source/features/latest-tag-button.tsx:99]: 5?

Comment (yakov116) [reply source/features/latest-tag-button.tsx:99]: Branch Name

File: source/features/default-branch-button.tsx (hunk 2 above)
Comment (yakov116) [reply source/features/default-branch-button.tsx:31]: Ok thought it would not hurt

Comment (yakov116) [reply source/features/default-branch-button.tsx:31]:
```suggestion
		url = pathParts.join('/')
```

File: source/features/default-branch-button.tsx [hunk 4]
```diff
@@ -17,11 +17,18 @@ async function init(): Promise<false | void> {
...
+		return false;
+	}
+
 	let url;
 	if (pageDetect.isRepoRoot()) {
 		url = `/${getRepoURL()}`;
 	} else {
-		url = replaceBranch(currentBranch, defaultBranch);
+		pathParts[5] = defaultBranch;
+		url = pathParts.join('/')
```
Comment (yakov116) [inline source/features/default-branch-button.tsx:31]:
```suggestion
		url = pathParts.join('/');
```

File: source/features/latest-tag-button.tsx (hunk 3 above)
Comment (yakov116) [reply source/features/latest-tag-button.tsx]: @fregante good catch. Not sure how I managed that!

File: source/features/default-branch-button.tsx [hunk 5]
```diff
@@ -17,11 +17,18 @@ async function init(): Promise<false | void> {
 		return false;
 	}
 
+	const pathParts = parseRoute(location.pathname);
+	// The branch selector will be on `isRepoCommitList()` **unless** you're in a folder/file
+	if (pageDetect.isRepoCommitList() && Boolean(pathParts[6])) {
```
Comment (fregante) [inline source/features/default-branch-button.tsx:22]: Boolean can be dropped now

**Comment by fregante** [comment]
Test it before asking for a review

File: source/features/default-branch-button.tsx [hunk 6]
```diff
@@ -17,11 +17,18 @@ async function init(): Promise<false | void> {
 		return false;
 	}
 
+	const pathParts = parseRoute(location.pathname);
+	// The branch selector will be on `isRepoCommitList()` **unless** you're in a folder/file
+	if (pageDetect.isRepoCommitList() && !pathParts[6]) {
```
Comment (fregante) [inline source/features/default-branch-button.tsx:22]: Isn’t this supposed to be 5?

**Comment by fregante** [comment]
There’s a change you can make to parseRoute to avoid these mistakes: have it export an object instead of an array. 

Bonus: add a `toString` function to that object to allow stringification without having the user call `join('/')`

**Comment by yakov116** [comment]
How does the , toString()} work I get a syntax error when I try

**Comment by fregante** [comment]
```js
{
	toString() {
		return `/${user}/${repo}/.............`
	}
}
```

**Comment by yakov116** [comment]
...Looks at the branch https://github.com/sindresorhus/refined-github/tree/parser...

@fregante I am in middle of testing. I want to just make sure that I did what you wanted before I continue.

File: source/libs/utils.ts [hunk 7]
```diff
@@ -156,23 +156,33 @@ export function getLatestVersionTag(tags: string[]): string {
...
 		user,
 		repository,
 		route,
-		currentBranch,
-		filePath
-	];
+		branch,
+		filePath,
+		toString() {
+			return `/${this.user}/${this.repository}/${this.route}/${this.branch}/${this.filePath}`.replace(/\/$/, '');
```
Comment (fregante) [inline source/libs/utils.ts:183]:
```suggestion
			return `/${user}/${repository}/${route}/${branch}/${filePath}`.replace(/\/$/, '');
```

**Comment by fregante** [comment]
That's right

File: source/libs/utils.ts (hunk 7 above)
Comment (yakov116) [reply source/libs/utils.ts:183]: When I left out the `this` any replacement I made it did not change it in the toString()

**Comment by fregante** [comment]
You can take this opportunity to move this function to its own file

```
source/github-helpers/parse-route.ts
```

**Comment by yakov116** [comment]
What a messy merge

File: source/features/deep-reblame.tsx [hunk 8]
```diff
@@ -66,17 +67,17 @@ async function redirectToBlameCommit(event: delegate.Event<MouseEvent, HTMLAncho
...
-	const pathnameParts = parseRoute(location.pathname);
-	const currentFilename = pathnameParts[5];
+	const path = parseRoute(location.pathname);
+	const {filePath: currentFilename} = path;
 
 	const spinner = <LoadingIcon className="mr-2"/>;
 	blameElement.firstElementChild!.replaceWith(spinner);
 
 	try {
 		const prBlameCommit = await getPullRequestBlameCommit(prCommit, prNumber, currentFilename);
```
Comment (fregante) [inline source/features/deep-reblame.tsx:77]:
```suggestion
		const prBlameCommit = await getPullRequestBlameCommit(prCommit, prNumber, path.filePath);
```

File: source/features/deep-reblame.tsx [hunk 9]
```diff
@@ -66,17 +67,17 @@ async function redirectToBlameCommit(event: delegate.Event<MouseEvent, HTMLAncho
 	const blameHunk = blameElement.closest('.blame-hunk')!;
 	const prNumber = looseParseInt(select('.issue-link', blameHunk)!.textContent!);
 	const prCommit = select<HTMLAnchorElement>('a.message', blameHunk)!.pathname.split('/').pop()!;
-	const pathnameParts = parseRoute(location.pathname);
-	const currentFilename = pathnameParts[5];
+	const path = parseRoute(location.pathname);
+	const {filePath: currentFilename} = path;
```
Comment (fregante) [inline source/features/deep-reblame.tsx:71]:
```suggestion
```

File: source/libs/utils.ts (hunk 7 above)
Comment (fregante) [reply source/libs/utils.ts]: [](https://knowyourmeme.com/memes/la-noire-doubt-press-x-to-doubt)

File: source/features/latest-tag-button.tsx [hunk 10]
```diff
@@ -95,7 +95,9 @@ async function init(): Promise<false | void> {
 	if (pageDetect.isRepoRoot()) {
 		href = `/${getRepoURL()}/tree/${latestTag}`;
 	} else {
-		href = replaceBranch(currentBranch, latestTag);
+		const pathnameParts = parseRoute(location.pathname);
+		pathnameParts.branch = latestTag;
+		href = pathnameParts.toString();
```
Comment (fregante) [inline source/features/latest-tag-button.tsx]: What about this pattern for assignment? (Only if it's done all in one place; don't refactor other code that has conditions, etc)
```suggestion
		href = {
			...parseRoute(location.pathname),
			branch: latestTag
		}.toString();
```

**Comment by yakov116** [comment]
@fregante that review came in very funny.

I cant apply any suggestions

File: source/libs/utils.ts (hunk 7 above)
Comment (yakov116) [reply source/libs/utils.ts]: I tried it again.
Without the this it does not replace anything when calling toString()

Comment (fregante) [reply source/libs/utils.ts]: Heh you’re right, the object is changed but the constants aren’t. I wasn’t thinking about the replacements

Comment (yakov116) [reply source/libs/utils.ts]: > Heh you’re right

It can happen... sometimes :smile:

**Comment by yakov116** [comment]
@fregante I tested this out seems to be fully working.

Side note: Deep reblame will sometimes fail if there is an open PR. I am not sure fully why, but there can be more than one PR associated with the commit.

See all the blame cases in https://github.com/sindresorhus/refined-github/pull/3083.

I have a way to fix it. But I will leave it for another pr.

**Pull request merged**